### PR TITLE
[RLlib] Implicit Q-Learning.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?
Implicit Q-Learning shows impressive results in experiments and its performance beats algorithms like CQL. However, up to now this algorithm has no implementation in RLlib.

This PR introduces Implicit Q-Learning (IQL) to RLlib. It comes with implementations of:
- The `DefaultIQLRLModule` and `DefaultIQLTorchRLModule` derived from the `DefaultSACRLModule`. 
- The `IQLLearner` and `IQLTorchLearner`, derived from the `SACLearner`.
- The `IQL`(`Algorithm`) derived from the `MARWIL`(`Algorithm`) and the `IQLConfig` derived from the `SACConfig`.
- A tuned example training SAC on our enormous `Pendulum-v1` data.

>__Note: __ The original IQL implementation updates the value network, actor network and Q-network sequentially during a single iteration. RLlib is dependend on TorchDDP for multi-gpu training and therefore it needs a single backward pass for all networks for the gradient synching in TorchDDP to work. As a result we can observe small cycles in the loss curves of all networks.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/iql/iql.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        # Run this from the ray directory root.
+        config.offline_data(
+            input_="./rllib/tests/data/pendulum-v1_enormous"
+        )
+        # Set the config object's env, used for evaluation.
+        config.environment(env="Pendulum-v1")
+        # Use to_dict() to get the old-style python config dict
+        # when running with tune.
+        tune.Tuner(
+            "IQL",
```
Comment (sven1977) [inline rllib/algorithms/iql/iql.py:55]: nit: I think for this, we'll have to add IQL to the algo registry.

`rllib.algorithms.registry.py`.

File: rllib/tuned_examples/iql/pendulum_iql.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    )
+    .rl_module(
+        model_config=DefaultModelConfig(
+            fcnet_hiddens=[256, 256],
+        ),
+    )
+)
+
+stop = {
+    f"{EVALUATION_RESULTS}/{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}": -800.0,
```
Comment (sven1977) [inline rllib/tuned_examples/iql/pendulum_iql.py]: Any chance we can get this to a higher return value in 10min on the CI? If not, how long would it have to run to reach say -200?

For DreamerV3, which takes 15-20min to get to >-200, I created a nightly release test instead.

File: rllib/BUILD [hunk 3]
```diff
@@ -1095,6 +1095,55 @@ py_test(
 #    args = ["--as-test", "--num-learners=2", "--num-gpus-per-learner=1"]
 # )
 
+# IQL
+# Pendulum-v1 (enormous)
```
Comment (sven1977) [inline rllib/BUILD:1099]: large or enormous? :)

We shouldn't allow enormous tests on the CI. See my comment below on maybe making this a nightly release test instead (see DreamerV3 Pendulum).

File: rllib/algorithms/iql/iql_learner.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+VF_PREDS_NEXT = "vf_preds_next"
+VF_LOSS = "value_loss"
+
+
+class IQLLearner(DQNLearner):
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    @override(DQNLearner)
+    def build(self) -> None:
+        # Define the expectile parameter(s).
+        self.expectile: Dict[ModuleID, TensorType] = LambdaDefaultDict(
```
Comment (sven1977) [inline rllib/algorithms/iql/iql_learner.py:24]: Any chance we can make these calls after the `super().build()` below?

Super methods should always be called as early as possible for clarity.

File: rllib/algorithms/iql/iql.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py:55]: On it. Thanks for the hint @sven1977 !

File: rllib/tuned_examples/iql/pendulum_iql.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/tuned_examples/iql/pendulum_iql.py]: @sven1977 This is regrettably dependent on the quality of expert data as we cannot improve exploration like in online RL. I will play a bit with the learning rates, `beta` and `expectile` - but the ultimate limit is the data quality

File: rllib/BUILD (hunk 3 above)
Comment (simonsays1980) [reply rllib/BUILD:1099]: This is not the test size but the data size. The `large` data shows less quality than the enormous one ;)

File: rllib/algorithms/iql/iql_learner.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql_learner.py:24]: Let me check - maybe there are things that get overridden otherwise

File: rllib/algorithms/iql/iql_learner.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        # Define the temperature for the acotr advantage loss.
+        self.temperature: Dict[ModuleID, TensorType] = LambdaDefaultDict(
+            lambda module_id: self._get_tensor_variable(
+                # Note, we want to train with a certain expectile.
+                [self.config.get_config_for_module(module_id).beta],
+                trainable=False,
+            )
+        )
+
+        # Store loss tensors here temporarily inside the loss function for (exact)
```
Comment (sven1977) [inline rllib/algorithms/iql/iql_learner.py:41]: super nit: Should we move this into `IQLTorchLearner` instead? This logic may be torch-specific.

File: rllib/algorithms/iql/iql.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+        # How often the target network should be updated.
+        self.target_network_update_freq = 0
+        # The weight for Polyak averaging.
+        self.tau = 1.0
+
+        # __sphinx_doc_end__
+        # fmt: on
+
+    @override(AlgorithmConfig)
```
Comment (sven1977) [inline rllib/algorithms/iql/iql.py]: We subclass from `MARWILConfig`. -> `override(MARWILConfig)`

File: rllib/algorithms/iql/iql.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        input_action_space,
+        device=None,
+    ):
+        pipeline = super().build_learner_connector(
+            input_observation_space=input_observation_space,
+            input_action_space=input_action_space,
+            device=device,
+        )
+
+        # Remove unneeded connectors from the MARWIL connector pipeline.
```
Comment (sven1977) [inline rllib/algorithms/iql/iql.py:160]: Very cool!

File: rllib/algorithms/iql/iql.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+        return pipeline
+
+    @override(MARWILConfig)
+    def validate(self) -> None:
+        # Call super's validation method.
+        super().validate()
+
+        # Ensure hyperparameters are meaningful.
+        if self.beta == 0.0:
```
Comment (sven1977) [inline rllib/algorithms/iql/iql.py]: super nit: Should we check for beta > 0.0 then instead?

File: rllib/algorithms/iql/iql_learner.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        # consumption later by the compute gradients function.
+        # Keys=(module_id, optimizer_name), values=loss tensors (in-graph).
+        self._temp_losses = {}
+
+        # Build the `DQNLearner` (builds the target network).
+        super().build()
+
+    @override(DQNLearner)
+    def remove_module(self, module_id: ModuleID) -> None:
```
Comment (sven1977) [inline rllib/algorithms/iql/iql_learner.py:47]: What about `add_module`?

File: rllib/algorithms/iql/torch/default_iql_torch_rl_module.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class DefaultIQLTorchRLModule(DefaultSACTorchRLModule, DefaultIQLRLModule):
+
+    framework: str = "torch"
+
+    @override(DefaultSACTorchRLModule)
+    def _forward_train(self, batch: Dict, **kwargs) -> Dict[str, Any]:
+
+        if self.inference_only:
```
Comment (sven1977) [inline rllib/algorithms/iql/torch/default_iql_torch_rl_module.py]: dumb question. Should we move this logic into all RLModules' `_forward_train()` by putting this if-block into the super class?

File: rllib/algorithms/iql/torch/default_iql_torch_rl_module.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            )
+
+        # Compute values for the current observations.
+        output[Columns.VF_PREDS] = self.compute_values(batch_curr)
+        # The values of the next observations are needed for the critic loss.
+        output[VF_PREDS_NEXT] = self.compute_values(batch_next)
+
+        return output
+
+    @override(DefaultIQLRLModule)
```
Comment (sven1977) [inline rllib/algorithms/iql/torch/default_iql_torch_rl_module.py]: nit: This is subclassed from the `ValueFunctionAPI`, no? The method is not implemented in `DefaultLQLRLModule`.

File: rllib/algorithms/iql/torch/default_iql_torch_rl_module.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    def compute_values(
+        self,
+        batch: Dict[str, Any],
+        embeddings: Optional[Any] = None,
+    ) -> TensorType:
+        # If no embeddings are provided make a forward pass on the encoder.
```
Comment (sven1977) [inline rllib/algorithms/iql/torch/default_iql_torch_rl_module.py:71]: Nice!

File: rllib/tuned_examples/iql/pendulum_iql.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        # The number of iterations to be run per learner when in multi-learner
+        # mode in a single RLlib training iteration. Leave this to `None` to
+        # run an entire epoch on the dataset during a single RLlib training
+        # iteration. For single-learner mode, 1 is the only option.
+        dataset_num_iters_per_learner=1 if not args.num_learners else None,
+    )
+    .training(
+        # To increase learning speed with multiple learners,
+        # increase the learning rate correspondingly.
+        lr=0.0008 * (args.num_learners or 1) ** 0.5,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tuned_examples/iql/pendulum_iql.py]: True, we need to set the individual lrs.

We should set default lr to -1 and add a check in `validate_config` to see whether the user has set that `lr` value by accident.

File: rllib/algorithms/iql/iql.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py:55]: Done

File: rllib/algorithms/iql/iql.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Great catch!

Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Done

File: rllib/algorithms/iql/iql.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Again, great catch! Yes we should!

Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Done

File: rllib/algorithms/iql/iql.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py:55]: Done

File: rllib/algorithms/iql/iql_learner.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql_learner.py:41]: They are in `SACLearner` as well. I have no problem to move them over, but then we have to implement the `setup` method also for the `IQLTorchLearner` (I think the `expectile` and `temperature` could be defined in there, too - we might not even need `get_default_tensor` anymore b/c we are on Torch anyways - only if we are thinking about JAX or so in the future it makes sense to keep it).

File: rllib/algorithms/iql/iql.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Great catch!! And done :)

File: rllib/algorithms/iql/iql.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql.py]: Great catch! Makes totally sense haha

File: rllib/algorithms/iql/iql_learner.py (hunk 9 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/iql_learner.py:47]: Great catch. Missing also in `SACLearner`. Will add.

Comment (simonsays1980) [reply rllib/algorithms/iql/iql_learner.py:47]: Done

File: rllib/algorithms/iql/torch/default_iql_torch_rl_module.py (hunk 11 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/torch/default_iql_torch_rl_module.py]: Another great catch! Done

Comment (simonsays1980) [reply rllib/algorithms/iql/torch/default_iql_torch_rl_module.py]: Great catch!

File: rllib/tuned_examples/iql/pendulum_iql.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/tuned_examples/iql/pendulum_iql.py]: Done.

File: rllib/algorithms/iql/torch/default_iql_torch_rl_module.py (hunk 10 above)
Comment (simonsays1980) [reply rllib/algorithms/iql/torch/default_iql_torch_rl_module.py]: Great catch @sven1977 ! This is actually already implemented in `RLModule.forward_train`, we just have to call `super`

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks @simonsays1980 !
